File: solutions/1562.py
def mod_add(a: int, b: int) -> int:
    return (a + b) % 1000000000

cache = {}

ALL_USED = (1 << 10) - 1

def starts_with(start: int, length: int, used) -> int:
    used = used | (1 << start)

    if length == 1:
        #print(start, length, used)
        return 1 if used == ALL_USED else 0

    if (start, length, used) in cache:
        return cache[ (start, length, used) ]

    result = 0
    if start > 0:
        result = mod_add(result, starts_with(start - 1, length - 1, used))

    if start < 9:
        result = mod_add(result, starts_with(start + 1, length - 1, used))

    cache[ (start, length, used) ] = result
    
    return result

def main(length):
    result = 0

    for i in range(1, 10):
        result = mod_add(result, starts_with(i, length, 0))

    return result
'''
a = 0
for i in range(1, 41):
    a += main(i)

print(a)
'''
print(main(int(input())))

File: solutions/1562.cpp
#include <cstdint>
#include <iostream>

constexpr int MOD = 1'000'000'000;
int cache[100][10][1 << 10];

int solve(uint len, uint8_t prev, uint16_t used) {
	if (len == 0) {
		return used == (1 << 10) - 1;
	}
	int& result = cache[len][prev][used];
	if (result != -1) {
		return result;
	}

	result = 0;

	if (0 < prev) {
		(result += solve(len - 1, prev - 1, used | (1 << (prev - 1)))) %= MOD;
	}
	if (prev < 9) {
		(result += solve(len - 1, prev + 1, used | (1 << (prev + 1)))) %= MOD;
	}
	return result;
}

int main() {
	uint len;
	std::cin >> len;

	for (uint i = 0; i < len; i++) {
		for (uint8_t prev = 0; prev < 10; prev++) {
			for (uint16_t used = 0; used < (1 << 10); used++) {
				cache[i][prev][used] = -1;
			}
		}
	}

	int result = 0;
	for (uint8_t ch = 1; ch <= 9; ch ++) {
		(result += solve(len - 1, ch, 1 << ch)) %= MOD;
	}

	std::cout << result << '\n';
}
